lms/djangoapps/enthought/auth.py

```diff
@@ -0,0 +1,169 @@
+            user = None
+
+        elif status == 'inactive':
+            user = User(username=username, is_active=False)
+
+        else:
+            try:
+                user = User.objects.get(username=username)
+
+            except User.DoesNotExist:
+                user = self._create_user(
+                    username   = username,
+                    password   = password,
+                    first_name = user_data['first_name'],
+                    last_name  = user_data['last_name'],
+                    is_active  = user_data['is_active']
+                )
+
+        return user
+
+    def get_user(self, user_id):
+        """ Return the User object for a user that has already been
+        authenticated by this backend.
+
+        """
+
+        try:
+            return User.objects.get(id=user_id)
+
+        except User.DoesNotExist:
+            return None
+
+    #### Private protocol #####################################################
+
+    _API_ROOT   = 'https://api.enthought.com'
+
+    def _authenticate(self, username, password):
+        """ Authenticate the user on api.enthought.com. """
+
+        url  = self._API_ROOT + '/token-auth/'
+        data = {'username': username, 'password': password}
+
+        authenticated = requests.post(url, data=data)
+        user_data     = self._get_user_data(username)
+
+        if authenticated.ok:
+            status = 'success'
+
+        else:
+            if user_data.get('is_active') is False:
+                status = 'inactive'
+
+            else:
+                status = 'invalid'
+
+        return status, user_data
+
+    def _create_user(self, username, password, first_name, last_name, is_active):
+        """ Create a user given username, password and other data.
+
+        Note that we have to do a full-fledged account creation with profile
+        and registration as well, so a simple `User.objects.create(...)` will
+        not work here. Instead, we call the edX supplied method to create an
+        account.
+
+        """
+
+        from student.views import _do_create_account
+
+        user, profile, registration = _do_create_account(
+            EnthoughtAccountCreationForm(
+                data         = {
+                    'username': username,
+                    'email'   : username,
+                    'password': password,
+                    'name'    : first_name + ' ' + last_name
+                },
+                tos_required = False
+            )
+        )
+
+        if is_active:
+            registration.activate()
+            registration.save()
+
+        return user
+
+    def _get_user_data(self, username):
+        """ Get user data from API. """
+
+        url = self._API_ROOT + '/api/users/%s/' % username
+        api_token = (
+            settings.ENTHOUGHT_API_TOKEN or
+            os.environ.get('ENTHOUGHT_API_TOKEN')
+        )
+
+        if api_token is None:
+            raise RuntimeError(
```

Comment:
I don't see how we can prevent this `RuntimeError` without running the code in production. Maybe it'd be better to use whatever value is there and handle the failed authentication in the standard way - by logging an exception and returning a message to the user - this will just blow up with a 500.